### PR TITLE
Various tags and mechanisms added

### PR DESCRIPTION
**Pull request by MusicScore**
- Adds `<context.source>` and `<context.source_type>` to `on entity combusts` event
- Adds `<w@world.world_type>` tag
- Adds `<e@entity.last_damage.max_duration>` tag and `dEntity.no_damage_duration` and `dEntity.max_no_damage_duration` mechanisms
- Allows `<e@entity.size>` tag and `dEntity.size` mechanism to work on Phantoms
- Adds `<e@entity.beam_target>` and `<e@entity.is_showing_bottom>` tags and `dEntity.beam_target` and `dEntity.is_showing_bottom` mechanisms
- Adds `<e@entity.spell>` tag and `dEntity.spell` mechanism

Fixes https://github.com/DenizenScript/Denizen-For-Bukkit/issues/1801
Fixes https://github.com/DenizenScript/Denizen-For-Bukkit/issues/1832
Fixes https://github.com/DenizenScript/Denizen-For-Bukkit/issues/1836
Fixes https://github.com/DenizenScript/Denizen-For-Bukkit/issues/1828

### CODE REVIEW

File: plugin/src/main/java/net/aufdemrand/denizen/objects/properties/entity/EntitySpell.java [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+        return null;
+    }
+
+    @Override
+    public void adjust(Mechanism mechanism) {
+
+        // <--[mechanism]
+        // @object dEntity
+        // @name spell
+        // @input Element(Boolean)
```
Comment (mergu) [inline plugin/src/main/java/net/aufdemrand/denizen/objects/properties/entity/EntitySpell.java]: boolean?

File: plugin/src/main/java/net/aufdemrand/denizen/objects/properties/entity/EntitySpell.java [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+        // @description
+        // Sets the spell the entity should cast.
```
Comment (mergu) [inline plugin/src/main/java/net/aufdemrand/denizen/objects/properties/entity/EntitySpell.java]: best to provide a link to valid spells here too

File: plugin/src/main/java/net/aufdemrand/denizen/objects/properties/entity/EntitySpell.java [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+        // @tags
+        // <e@entity.spell>
+        // -->
+
+        if (mechanism.matches("spell") && mechanism.requireEnum(false, Spellcaster.Spell.values())) {
+            setSpell(Spellcaster.Spell.valueOf(mechanism.getValue().asString().toUpperCase()));
```
Comment (mergu) [inline plugin/src/main/java/net/aufdemrand/denizen/objects/properties/entity/EntitySpell.java]: this isn't a huge deal, but why did you move this into its own method? it's only called here and is a single line

File: plugin/src/main/java/net/aufdemrand/denizen/events/entity/EntityCombustsScriptEvent.java [hunk 4]
```diff
@@ -108,6 +113,20 @@ public dObject getContext(String name) {
         else if (name.equals("duration")) {
             return new Duration(burntime);
         }
+        else if (name.equals("source")) {
```
Comment (mergu) [inline plugin/src/main/java/net/aufdemrand/denizen/events/entity/EntityCombustsScriptEvent.java]: i would still like to see the CombustByBlock source here and a removal of the note in the docs

File: plugin/src/main/java/net/aufdemrand/denizen/objects/dEntity.java [hunk 5]
```diff
@@ -3175,6 +3226,48 @@ public void adjust(Mechanism mechanism) {
...
+        // -->
+        if (NMSHandler.getVersion().isAtLeast(NMSVersion.v1_9_R2) && mechanism.matches("collidable")
+                && mechanism.requireBoolean()) {
+            getLivingEntity().setCollidable(value.asBoolean());
+        }
+
+        // <--[mechanism]
+        // @object dEntity
+        // @name no_damage_duration
+        // @input Element(Boolean)
```
Comment (mergu) [inline plugin/src/main/java/net/aufdemrand/denizen/objects/dEntity.java]: boolean?

File: plugin/src/main/java/net/aufdemrand/denizen/objects/dEntity.java [hunk 6]
```diff
@@ -3175,6 +3226,48 @@ public void adjust(Mechanism mechanism) {
...
+        // <e@entity.last_damage.max_duration>
+        // -->
+        if (mechanism.matches("no_damage_duration") && mechanism.requireObject(Duration.class)) {
+            getLivingEntity().setNoDamageTicks(value.asType(Duration.class).getTicksAsInt());
+        }
+
+        // <--[mechanism]
+        // @object dEntity
+        // @name max_no_damage_duration
+        // @input Element(Boolean)
```
Comment (mergu) [inline plugin/src/main/java/net/aufdemrand/denizen/objects/dEntity.java]: boolean?

File: plugin/src/main/java/net/aufdemrand/denizen/events/entity/EntityCombustsScriptEvent.java (hunk 4 above)
Comment (mergu) [reply plugin/src/main/java/net/aufdemrand/denizen/events/entity/EntityCombustsScriptEvent.java]: We talked about this in discord, but just to note here - This always returning null is a bug that may be fixed in the future. Because of this, we should preemptively implement it and just note in the docs that it currently always returns null.
https://hub.spigotmc.org/stash/projects/SPIGOT/repos/craftbukkit/browse/nms-patches/Entity.patch#198
https://hub.spigotmc.org/stash/projects/SPIGOT/repos/craftbukkit/browse/nms-patches/Entity.patch#265

**Comment by mcmonkey4eva** [comment]
What's the current status of this PR? In terms of testing / stability of each of the things in it

**Comment by MusicScore** [comment]
Currently, I believe I have fully tested the PR. Every tag and mechanism was checked to see if they were working and if they took appropriate input across all relevant versions (1.8.8, 1.9.4, 1.11.2, 1.12.2, and 1.13.2).

There were a few minor things, like the collidable meta note I added and the swimming mec not quite working on players (but seeming to work fine on the Drowned in regards to adjusting them and later seeing `true` being returned from `<e@entity.swimming>` when the Drowned swam about underwater). To get players to swim underwater, it seems that you need to adjust `sprinting` to true rather than `swimming`.

**Comment by Elxafil** [comment]
When will this be merged?

**Comment by mcmonkey4eva** [comment]
Never. Christmas is cancelled.
